src/Cryptol/TypeCheck/Infer.hs

```diff
@@ -1019,54 +1035,202 @@ checkSigB b (Forall as asmps0 t0, validSchema) = case thing (P.bDef b) of
+              checkPropGuardCase (guards0, e0) = do
+                mb_rng <- case P.bSignature b of
+                  Just (P.Forall _ _ _ mb_rng) -> pure mb_rng
+                  _ -> panic "checkSigB"
+                    [ "Used constraint guards without a signature, dumbwit, at "
+                    , show . pp $ P.bName b ]
+                -- check guards
+                (guards1, goals) <-
+                  second concat . unzip <$>
+                  checkPropGuard mb_rng `traverse` guards0
+                -- try to prove all goals
+                su <- proveImplication True (Just name) as (asmps1 <> guards1) goals
+                extendSubst su
+                -- Preprends the goals to the constraints, because these must be
+                -- checked first before the rest of the constraints (during
+                -- evaluation) to ensure well-definedness, since some
+                -- constraints make use of partial functions e.g. `a - b`
+                -- requires `a >= b`.
+                let guards2 = (goal <$> goals) <> concatMap pSplitAnd (apSubst su guards1)
+                (_t, guards3, e1) <- checkBindDefExpr asmps1 guards2 e0
+                e2 <- applySubst e1
+                pure (guards3, e2)
+
+          cases1 <- mapM checkPropGuardCase cases0
+
+          -- Here, `props` is a conjunction of multiple constraints, and so the
+          -- negation is the disjunction of the negation of each conjunct i.e.
+          -- "not (A and B and C) <=> (not A) or (not B) or (not C)". This is
+          -- one of DeMorgan's laws of boolean logic. Since there are no
+          -- constraint disjunctions, we encode a disjunction of conjunctions as
+          -- `[[Prop]]`. For exhaustive checking, each disjunct will need to be
+          -- checked independently, and all must pass in order to be considered
+          -- exhaustive.
+
+          let negateSimpleNumProp :: Prop -> Maybe [Prop]
+              negateSimpleNumProp prop = case prop of
+                TCon tcon tys -> case tcon of
+                  PC pc -> case pc of
+                    -- not x == y  <=>  x /= y
+                    PEqual -> pure [TCon (PC PNeq) tys]
+                    -- not x /= y  <=>  x == y
+                    PNeq -> pure [TCon (PC PEqual) tys]
+                    -- not x >= y  <=>  x /= y and y >= x
+                    PGeq -> pure [TCon (PC PNeq) tys, TCon (PC PGeq) (reverse tys)]
+                    -- not fin x  <=>  x == Inf
+                    PFin | [ty] <- tys -> pure [TCon (PC PEqual) [ty, tInf]]
+                         | otherwise -> panicInvalid
+                    -- not True  <=>  0 == 1
+                    PTrue -> pure [TCon (PC PEqual) [tZero, tOne]]
+                    _ -> mempty
+                  TC _tc -> panicInvalid
+                  TF _tf -> panicInvalid
+                  TError _ki -> panicInvalid -- TODO: where does this come from?
```

Comment:
`TError` results from an "impossible" term (e.g., `1-2`).  It acts kind of like an exception, and when you encounter it you should just propagate it (i.e., `pNotNumeric (TError x) = TError x`).